### ISSUE TITLE
Number of processes increases every time remote window is reloaded

### ISSUE THREAD

**Issue by jcready** [microsoft/vscode#211462]
Does this issue occur when all extensions are disabled?: Yes

Version: 1.88.1 (Universal)
Commit: e170252f762678dec6ca2cc69aba1570769a5d39
Date: 2024-04-10T17:42:52.765Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Darwin x64 23.4.0

Steps to Reproduce:

1. Open remote workspace with a TypeScript file
2. Wait for the "Initializing JS/TS language features" status bar indicator to stop appearing.
3. Run this script to get the list of ionotify watcher processes: https://gist.github.com/richtera/f6c228d0a28f69a213aa7afc38a1c98d
4. Note the number of processes whose `COMMAND` looks like `~/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node`
5. Reload the window.
6. Repeat steps 2 through 4 and see that the number noted from step 4 increased from the last time.

First run:
```
   INOTIFY   INSTANCES
   WATCHES      PER   
    COUNT     PROCESS   PID USER         COMMAND
------------------------------------------------------------
   34318         1      794015 jcready     /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/out/bootstrap-fork --type=fileWatcher
   14248         1      809038 jcready     /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node --max-old-space-size=3072 /home/jcready/repo/node_modules/typescript/lib/tsserver.js --useInferredProjectPerProjectRoot --enableTelemetry --cancellationPipeName /tmp/vscode-typescript1551/48d5
```

After reloading the window:
```
   INOTIFY   INSTANCES
   WATCHES      PER   
    COUNT     PROCESS   PID USER         COMMAND
------------------------------------------------------------
   34318         1      815144 jcready     /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/out/bootstrap-fork --type=fileWatcher
   34318         1      794015 jcready     /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/out/bootstrap-fork --type=fileWatcher
   14248         1      814816 jcready     /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node --max-old-space-size=3072 /home/jcready/repo/node_modules/typescript/lib/tsserver.js --useInferredProjectPerProjectRoot --enableTelemetry --cancellationPipeName /tmp/vscode-typescript1551/80e6
   14248         1      809038 jcready     /home/jcready/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/node --max-old-space-size=3072 /home/jcready/repo/node_modules/typescript/lib/tsserver.js --useInferredProjectPerProjectRoot --enableTelemetry --cancellationPipeName /tmp/vscode-typescript1551/48d5
```

**Comment by bpasero** [comment]
This does not seem file watcher process specific as the other process `typescript/lib/tsserver.js` is also increasing each time. Moving to some remote folks for chiming in, I am not sure how sessions are tracked and processes recycled.

**Comment by aeschli** [comment]
Nice script
I tested on WSL and all looks fine there. No increase of watchers after a reload.

**Comment by bpasero** [comment]
@jcready what kind of remote is this? SSH? Docker? WSL? CLI?

**Comment by jcready** [comment]
SSH connection to AWS `Linux x64 6.2.0-1010-aws` specifically running Ubuntu 22.04.

**Comment by github-yxb** [comment]
Same issue.

Version: 1.88.1
Commit: e170252f762678dec6ca2cc69aba1570769a5d39
Date: 2024-04-10T17:42:52.765Z (2 wks ago)
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: Darwin x64 23.4.0

remote ssh on:
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.3 LTS
Release:        22.04
Codename:       jammy


the scripts output:
![image](https://github.com/microsoft/vscode/assets/7059300/ffd1dcc0-5cfc-4ca9-a05b-94c5b17d0706)

**Comment by bpasero** [comment]
I can reproduce that after window reload extension host and file watcher processes pile up when connected to a SSH remote:

![image](https://github.com/microsoft/vscode/assets/900690/44a2c3e3-6037-47fe-9eff-03beb687cc03)

**Comment by bpasero** [comment]
One interesting observation is that the PID of "remote agent" seems to stay the same, indicating that this process is kept alive and thus child processes such as file watcher and extension host do not just terminate.

Btw this reproduces for older versions, so its not a regression.

**Comment by alexdima** [comment]
@connor4312 

We have a number of mechanisms in place for dealing with this scenario and it looks like all are failing. See below for what 1, 2 and 3 are.

In a perfect world, we would have sufficient time on the client machine to be able to forward the final messages sent via 1. and 2. from the renderer process. But that doesn't always work (sometimes the renderer crashes), so at least 3. would be important to get right (i.e. closing the sockets to the server when the renderer reloads / the WebSocket is closed). We would preferably get all 3 working OK like they work in the case without a resolver.

By the way, to try out core remoting support without a resolver you can use (I just checked again and 1. and 2. work without a resolver):
```
./scripts/code-server.sh --without-connection-token
./scripts/code.sh --folder-uri=vscode-remote://localhost:9888/Users/alex/src/working_dir
```

### 1. Sending `ProtocolMessageType.Disconnect` on the management connection

The call stack looks something like this:
* [lifecycleService.onDidShutdown](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/services/extensions/common/abstractExtensionService.ts#L198-L206)
* `RemoteAgentConnection.dispose`
* `PersistentConnection.dispose`
* `Client<TContext = string>.dispose`
* [`PersistentProtocol.sendDisconnect`](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/base/parts/ipc/common/ipc.net.ts#L600)
* The buffered write is flushed so `WebSocket.send` should be called immediately and should "make it" before the window goes down.

### 2. Sending a `Terminate` and a `ProtocolMessageType.Disconnect` on the remote extension host connection
* [lifecycleService.onDidShutdown](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/services/extensions/common/abstractExtensionService.ts#L198-L206)
* [_doStopExtensionHosts](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/services/extensions/common/abstractExtensionService.ts#L677-L682) going first to the remote extension host before the local one
* [ExtensionHostManager.dispose](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/services/extensions/common/extensionHostManager.ts#L164)
* [sending Terminate](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/services/extensions/common/remoteExtensionHost.ts#L277), which is a concept extension hosts have irrespective of remote
* [`PersistentProtocol.sendDisconnect`](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/services/extensions/common/remoteExtensionHost.ts#L278)
* The buffered write is flushed so `WebSocket.send` should be called immediately and should "make it" before the window goes down.


### 3. Listening to socket close events on the server side
On the server side we listen to socket close events and then install a 3hrs timeout to wait for reconnections.
* For the management connection, this happens [here](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/server/node/remoteExtensionManagement.ts#L74-L78). As soon as this fires, we log as INFO "The client has disconnected, will wait for reconnection...". So seeing this log is a good indication that we're getting a socket close event. See below in the server log, this log message doesn't appear.
* For the remote extension host, this happens [here](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/workbench/api/node/extensionHostProcess.ts#L199-L203).
* 3hrs is a lot, so to support cases in which 1. & 2. failed, whenever a new succesful connection occurs (like in this repro with a window reload), we loop through all connections, and for those who are on the 3hrs wait time, we shorten their wait time to 5min. This happens [here](https://github.com/microsoft/vscode/blob/3982027f02dee5aa52622f7f2812c70f04ce337a/src/vs/server/node/remoteExtensionHostAgentServer.ts#L397-L407).

### Server log
This is when reloading around 3 times. Note there is no "The client has disconnected, will wait for reconnection..." message.
```
2024-05-03 11:31:55.228 [info] 




2024-05-03 11:31:55.386 [info] Extension host agent started.
2024-05-03 11:31:55.446 [info] [<unknown>][c0889be6][ManagementConnection] New connection established.
2024-05-03 11:31:55.446 [info] [<unknown>][1bc96464][ExtensionHostConnection] New connection established.
2024-05-03 11:31:55.571 [info] [<unknown>][1bc96464][ExtensionHostConnection] <83418> Launched Extension Host Process.
2024-05-03 11:31:55.630 [warning] Skipping extension /Users/alex/.vscode-server-insiders/extensions/ms-vscode.js-debug-1.72.1 in favour of the builtin extension /Users/alex/.vscode-server-insiders/cli/servers/Insiders-a1f2ea3162c704a5890c35dde4df6ee46283d957/server/extensions/ms-vscode.js-debug.
2024-05-03 11:36:55.391 [info] New EH opened, aborting shutdown
2024-05-03 11:37:11.365 [info] [<unknown>][c5315465][ExtensionHostConnection] New connection established.
2024-05-03 11:37:11.366 [info] [<unknown>][d37d0172][ManagementConnection] New connection established.
2024-05-03 11:37:11.368 [info] [<unknown>][c5315465][ExtensionHostConnection] <10544> Launched Extension Host Process.
2024-05-03 11:37:11.428 [warning] Skipping extension /Users/alex/.vscode-server-insiders/extensions/ms-vscode.js-debug-1.72.1 in favour of the builtin extension /Users/alex/.vscode-server-insiders/cli/servers/Insiders-a1f2ea3162c704a5890c35dde4df6ee46283d957/server/extensions/ms-vscode.js-debug.
2024-05-03 11:42:11.822 [info] [<unknown>][1bc96464][ExtensionHostConnection] <83418> Extension Host Process exited with code: 0, signal: null.
2024-05-03 11:58:27.323 [info] [<unknown>][d7222e53][ManagementConnection] New connection established.
2024-05-03 11:58:27.325 [info] [<unknown>][e4939ad8][ExtensionHostConnection] New connection established.
2024-05-03 11:58:27.330 [info] [<unknown>][e4939ad8][ExtensionHostConnection] <21783> Launched Extension Host Process.
2024-05-03 11:58:27.358 [warning] Skipping extension /Users/alex/.vscode-server-insiders/extensions/ms-vscode.js-debug-1.72.1 in favour of the builtin extension /Users/alex/.vscode-server-insiders/cli/servers/Insiders-a1f2ea3162c704a5890c35dde4df6ee46283d957/server/extensions/ms-vscode.js-debug.
2024-05-03 12:03:27.691 [info] [<unknown>][c5315465][ExtensionHostConnection] <10544> Extension Host Process exited with code: 0, signal: null.
```

**Comment by connor4312** [comment]
> The buffered write is flushed so WebSocket.send should be called immediately and should "make it" before the window goes down.

I don't think this is the case, at least not for node: the underlying implementations of flush is asynchronous ([1](https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/src/vs/base/parts/ipc/common/ipc.net.ts#L448) -> [2](https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/src/vs/base/parts/ipc/common/ipc.net.ts#L520) -> [3](https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/src/vs/base/parts/ipc/node/ipc.net.ts#L122)) and the window may terminate before they finish.

> As soon as this fires, we log as INFO "The client has disconnected, will wait for reconnection...". So seeing this log is a good indication that we're getting a socket close event. See below in the server log, this log message doesn't appear.

In the SSH case, we used a unix socket / named pipe to connect to the server. I observe that when the remote goes away the socket gets an `end` event but no `close` event. I'm not sure why this would happen and have trouble replicating this behavior to happen in any simpler program 😕 From the docs that seems to indicate the remote side finished its transmission and the local side should finish up, but that never happens...

On the Rust side we do properly await the shutdown

https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/cli/src/tunnels/server_bridge.rs#L60

**Comment by alexdima** [comment]
> I don't think this is the case, at least not for node: the underlying implementations of flush is asynchronous ([1](https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/src/vs/base/parts/ipc/common/ipc.net.ts#L448) -> [2](https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/src/vs/base/parts/ipc/common/ipc.net.ts#L520) -> [3](https://github.com/microsoft/vscode/blob/d68eddc5cc0e277df96a1748742c7f6b7f183328/src/vs/base/parts/ipc/node/ipc.net.ts#L122)) and the window may terminate before they finish.

:+1: Yes, that's right. But we don't use nodejs sockets, we use [WebSockets](https://github.com/microsoft/vscode/blob/886f3b7229708450cb15325cde2075dd0a9496fe/src/vs/platform/remote/browser/browserSocketFactory.ts#L50) from the renderer (via Chromium). That is why the server gets the `ProtocolMessageType.Disconnect` message in the vanilla (no resolver) case:

```
./scripts/code-server.sh --without-connection-token
./scripts/code.sh --folder-uri=vscode-remote://localhost:9888/Users/alex/src/working_dir
```

> I observe that when the remote goes away the socket gets an end event but no close event.

I'm not entirely sure what you mean here. But there is one design flaw (flaw is a too-strong word here) of Unix that I'm aware of in this area, namely it isn't possible to really create a tunnel that forwards all errors, like ETIMEOUT for example. In a tunnel implementation, there is usually a local socket and a remote socket. When the remote socket closes due to an ETIMEOUT, that usually means the OS timed out and decided to close to socket, meaning the TCP/IP socket closed without a FIN. For a tunnel implementor, if you get such an error from a remote socket, you cannot close the local socket with the same error. The best the UNIX APIs allow you to do is to close the local socket, which will cause it to get closed with a FIN. Technically, this is not 100% reflecting what the remote socket did, but in most cases it should be fine. Are you hitting something like this?

**Comment by connor4312** [comment]
I think so, I see [this](https://github.com/microsoft/vscode/blob/eccc47493a39ad1e6ecf9b754989d3744759a10f/src/vs/base/parts/ipc/node/ipc.net.ts#L61-L62) fire but not [this](https://github.com/microsoft/vscode/blob/eccc47493a39ad1e6ecf9b754989d3744759a10f/src/vs/base/parts/ipc/node/ipc.net.ts#L55-L56)

**Comment by frewsxcv** [comment]
I'm also running into this issue (https://github.com/microsoft/vscode/issues/214605), and feel like this is going to come up more as long as `typescript.tsserver.experimental.useVsCodeWatcher` continues to be rolled out

**Comment by mlavin** [comment]
I've been debugging this issue which is causing high memory/CPU issues within some shared hosts at $JOB. Walking back my VS Code version locally, the most recent version where I **don't** see this issue is:

Version: 1.87.2 (user setup)
Commit: 863d2581ecda6849923a2118d93a088b0745d9d6
Date: 2024-03-08T15:20:17.278Z
Electron: 27.3.2
ElectronBuildId: 26836302
Chromium: 118.0.5993.159
Node.js: 18.17.1
V8: 11.8.172.18-electron.0
OS: Windows_NT x64 10.0.22631

Using 1.87.2, I see the server log after `F1 > Reload Window` as:
```
2024-06-07 11:32:40.347 [info] [127.0.0.1][c8882321][ManagementConnection] The client has disconnected gracefully, so the connection will be disposed.
2024-06-07 11:32:40.368 [info] [127.0.0.1][05af74ca][ExtensionHostConnection] <1504397> Extension Host Process exited with code: 0, signal: null.
2024-06-07 11:32:40.368 [info] Last EH closed, waiting before shutting down
2024-06-07 11:32:44.844 [info] Got delay-shutdown request while in shutdown timeout, delaying
2024-06-07 11:32:44.844 [info] Cancelling previous shutdown timeout
2024-06-07 11:32:45.195 [info] [127.0.0.1][b1ad1f38][ManagementConnection] New connection established.
2024-06-07 11:32:45.354 [info] [127.0.0.1][8aeae564][ExtensionHostConnection] New connection established.
2024-06-07 11:32:45.361 [info] [127.0.0.1][8aeae564][ExtensionHostConnection] <1505085> Launched Extension Host Process.
```
as expected/described in part 3 of https://github.com/microsoft/vscode/issues/211462#issuecomment-2092773400

**Comment by connor4312** [comment]
I finally found an isolated way to reproduce it and chasing geese for a while. The secret is that you need to be an Node.js _HTTP_ server listening on a unix domain socket, which is what we do for our connections even though we treat them as TCP-ish. I previously was using a direct `net.server` which did not repro the issue.

```js
const http = require('http');

const socketPath = '/tmp/socket.sock';
const httpServer = http.createServer(function () {});

httpServer.on('upgrade', async (req, socket) => {
	console.log('---');
	console.log('Client connected');

	socket.on('end', () => {
		console.log('Client end');
	});
	socket.on('close', () => {
		console.log('Client close');
	});
	socket.on('data', (data) => {
		console.log('Client sent:', data.toString());
	});
});
```

Then run

```
curl --include \
     --no-buffer \
     --header "Connection: Upgrade" \
     --header "Upgrade: websocket" \
     --header "Host: example.com:80" \
     --header "Origin: [http://example.com:80](http://example.com/)" \
     --header "Sec-WebSocket-Key: SGVsbG8sIHdvcmxkIQ==" \
     --header "Sec-WebSocket-Version: 13" \
     --unix-socket /tmp/socket.sock\
     http://example.com:80/
```

and ctrl+c. Notice you see the same behavior of an 'end' event without ever getting a 'close' event. This seems like a niche Node.js bug.

I think an okay workaround for now is just adding a 30 second timeout to `socket.destroy()` in the 'end' listener. That does then result in the 'close' event being fired, and 30 seconds should be ample time for any network buffers to drain. I'll also verify this repro's on Node 22 and open an upstream bug.

**Comment by jcready** [comment]
Wow, that is an interesting bug. Thank you for fixing the issue! I am curious though, based on [the comment just above yours](#issuecomment-2155386167), it seems like this bug only showed up between:

Version: 1.87.2
Commit: https://github.com/microsoft/vscode/commit/863d2581ecda6849923a2118d93a088b0745d9d6
Date: 2024-03-08T15:20:17.278Z
Electron: 27.3.2
ElectronBuildId: 26836302
Chromium: 118.0.5993.159
Node.js: 18.17.1
V8: 11.8.172.18-electron.0

and 

Version: 1.88.1
Commit: https://github.com/microsoft/vscode/commit/e170252f762678dec6ca2cc69aba1570769a5d39
Date: 2024-04-10T17:42:52.765Z
Electron: 28.2.8
ElectronBuildId: 27744544
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0

Which is quite a narrow range. So do we have any clue about what commit introduced this? Was it actually some code change or was it due to the version bumps in Electron, Chromium, Node.js, or V8?

**Comment by connor4312** [comment]
Yes, there was an underlying change which switched the connection to the VS Code server from happening over a TCP connection to happening over a unix domain socket / windows named pipe.

**Comment by RobbyCBennett** [comment]
I've tested this on the latest Insiders build and it's still running on the server after closing all Insiders GUI windows.

Things I checked:
- I waited more than 30 seconds.
- The commit of my build (0a417782b1781d39803b8f1791cfb22f8835ddae) is definitely after the commit of the fix (dc9701399307b9e4919116c82dd30eda86deccc2).

Client:
```
Version: 1.91.0-insider
Commit: 0a417782b1781d39803b8f1791cfb22f8835ddae
Date: 2024-06-13T05:48:47.964Z
Electron: 29.4.0
ElectronBuildId: 9593362
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Windows_NT x64 10.0.22631
```

Host:
```
Architecture: x86_64
OS: Ubuntu 22.04
```

**Comment by connor4312** [comment]
Yes, the server is still going to be running, but it should not proliferate indefinitely as it did before. Based on Alex's comment above, this should now happen:

> On the server side we listen to socket close events and then install a 3hrs timeout to wait for reconnections... 3hrs is a lot, so to support cases in which 1. & 2. failed, whenever a new succesful connection occurs (like in this repro with a window reload), we loop through all connections, and for those who are on the 3hrs wait time, we shorten their wait time to 5min.

**Comment by alexdima** [comment]
@connor4312 Did you ever figure out why these `ProtocolMessageType.Disconnect` messages don't make it over? Could you figure out if they at least make it over to the local resolver extension? Is there something I can do to help with getting these messages flushed from the renderer before it goes down?

**Comment by RobbyCBennett** [comment]
This solves this issue of having multiple instances, but there is still the issue that 1 instance remains running on the server. This is bad for VS Code extensions that use restrictive licensing, for example if a license only allows 1 instance for the user. If an extension thinks that it is still being used after closing the window, then the user won't be able to use it properly.

**Comment by Jokeswar** [comment]
As @RobbyCBennett said, any extension that uses a licensing system will suffer from license starvation. As the remote extension host is still kept alive, albeit for 5 minutes instead of 3 hours, users might run into problems where they cannot acquire a license (in cases where they use a licensing management system), because some remote extension host is running as a "zombie" process on a machine that no one is connecting to. This is a breaking change, and for the worse, considering that a few releases ago everything was working properly! I hope there is a plan to implement the other 2 methods @connor4312 mentioned, so at least when VS Code is closed gracefully the remote server is properly notified!

**Comment by connor4312** [comment]
> Did you ever figure out why these ProtocolMessageType.Disconnect messages don't make it over? Could you figure out if they at least make it over to the local resolver extension? Is there something I can do to help with getting these messages flushed from the renderer before it goes down?

Ideally I think we could use a shutdown participant to make sure they get out (though I'm not sure if layering works there). The disconnect message is async (https://github.com/microsoft/vscode/issues/211462#issuecomment-2121154266) but because message passing is handled by the local extension host, that could also be terminating before the flush is propagated though.

**Comment by mlavin** [comment]
I tested this change in my environment on the latest insiders build

```
Version: 1.91.0-insider (user setup)
Commit: 0354163c1c66b950b0762364f5b4cd37937b624a
Date: 2024-06-26T10:11:25.363Z
Electron: 29.4.0
ElectronBuildId: 9728852
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Windows_NT x64 10.0.22631
```

using Remote - SSH Windows -> Linux host. At first, I was still seeing the proliferation of infinite file watchers on reload and the file watchers not being killed when VS Code is closed. The install location is an NFS mount (could be a red herring but there were some notes about hosts with issues with locking/distributed file systems) so I tried adjusting some of the extension options:

- Toggled `Remote.SSH: Lockfiles In Tmp` - no effect
- Toggled `Remote.SSH: Use Flock` - no effect
- Toggled `Remote.SSH: Remote Server Listen On Socket` - no effect

Finally turning `Remote.SSH: Use Exec Server` off (other options using the defaults) **did** address the problem for me. Maybe also worth noting that disabling this option also works for me in 1.90.2 to prevent the zombie file watchers.

**Comment by mlavin** [comment]
Yes I understand that's the expectation. I continue to see these processes beyond that timeout without disabling this option.

**Comment by connor4312** [comment]
Can you capture of a gif of what you're seeing? https://gifcap.dev/ is useful

**Comment by mlavin** [comment]
![Recording 2024-06-26 at 18 39 48](https://github.com/microsoft/vscode/assets/149057/e1ef6b49-f8a3-4664-ae17-df70f2faac47)

**Comment by mlavin** [comment]
As another data point, two of these watchers from the capture (PID 2291296 and 2291471) are still running ~30 mins later having closed VS Code. The first one did timeout.

**Comment by fcsonline** [comment]
We're also suffering this performance degradation that makes VSCode hang and also the ssh connection so unstable. We usually suffer this issue after performing `git` operations like `checkout`, `rebase` and others. After VSCode hang, and connecting with ssh directly to the remote machine, we can see that machine is relative calm and the `vscode-server` logs are printing for many minutes this messages:

```
2024-06-26 12:43:46.439 [warning] [File Watcher (node.js)] Watcher shutdown because watched path got deleted
2024-06-26 12:43:47.077 [warning] [File Watcher (node.js)] Watcher shutdown because watched path got deleted
2024-06-26 12:43:47.704 [warning] [File Watcher (node.js)] Watcher shutdown because watched path got deleted
2024-06-26 12:43:48.332 [warning] [File Watcher (node.js)] Watcher shutdown because watched path got deleted
2024-06-26 12:43:48.946 [warning] [File Watcher (node.js)] Watcher shutdown because watched path got deleted
2024-06-26 12:43:50.328 [info] [<unknown>][2856f39f][ManagementConnection] The client has reconnected.
2024-06-26 12:43:51.237 [error] CodeExpectedError: Could not find pty on pty host
    at A.U (/home/factorial/.vscode-server/cli/servers/Stable-5437499feb04f7a586f677b155b039bc2b3669eb/server/out/vs/platform/terminal/node/ptyHostMain.js:55:30556)
```

**Comment by alexdima** [comment]
@fcsonline I don't think the issue you are seeing is necessarily related to the issue discussed here. Doing fs operations via git is not the root cause of "leaking" processes in this thread.

We have heard that there may be problems with the file watcher which could lead to 100% CPU usage in case of large repositories, especially for TypeScript projects. If that is what you might be running into, but in any case actually, it would be better to create a new issue.

**Comment by connor4312** [comment]
I'm going to reopen this to track the remaining issue to actually disconnect gracefully. PR incoming in a minute, a little more involved...

**Comment by connor4312** [comment]
This is now fixed properly in the next VS Code Insiders and SSH prerelease extensions.

**Comment by RobbyCBennett** [comment]
I just tried this in the latest VS Code Insiders and I still see the issue. The server remains running after the client closes. The number of processes increases when I re-open it.

```
Version: 1.92.0-insider
Commit: e6a828c5957959b7a18de37a567d850880101641
Date: 2024-07-03T19:56:36.468Z
Electron: 29.4.0
ElectronBuildId: 9728852
Chromium: 122.0.6261.156
Node.js: 20.9.0
V8: 12.2.281.27-electron.0
OS: Windows_NT x64 10.0.22631
```

**Comment by connor4312** [comment]
A new version of SSH prerelease has not yet been published

**Comment by RobbyCBennett** [comment]
Sorry if I'm being annoying, but my SSH extension updated to a new pre-release (version v0.113.2024070816) in the latest VS Code Insiders (1.92.0-insider, 01eef174c20a0c9ce7d49db93d9dade6c684f66e, 2024-07-15T12:53:32.802Z) and I still see the issue. The process remains open and every close results in 1 more process.

The version numbers of the extension and editor indicate that they were released after all commits mentioned in this issue.

**Comment by connor4312** [comment]
Can you capture of a gif of what you're seeing? https://gifcap.dev/ is useful

**Comment by RobbyCBennett** [comment]
Sure.

https://github.com/RobbyCBennett/big-gif/blob/main/Recording%202024-07-15%20at%2016.32.38.gif

I had a problem recording a video with OBS with my new Intel GPU, so I made a GIF with gifcap.dev. The GIF was too big for a GitHub comment, so I made a big-gifs repo instead.

**Comment by connor4312** [comment]
It looks like in your case the SSH connection script did not 'realize' it's exited yet. Usually this is killed when the SSH connection is closed, but sometimes it may not be. The check is done every 180 seconds, I would expect the processes you see to clean up after that time.

**Comment by RobbyCBennett** [comment]
No, they don't clean up after 180 seconds. Ideally, it should exit immediately after the client does, not after 180 seconds.

In the following screenshot, I closed the client at around 44 seconds and the server lasts seemingly forever.

![image](https://github.com/user-attachments/assets/f60ffea4-d469-4afb-8417-7e777d042b6f)

**Comment by yb3616** [comment]
![image](https://github.com/user-attachments/assets/8d74cb6d-ac00-46f3-b16d-c4d8a9f08eeb)

After 180sec

![image](https://github.com/user-attachments/assets/5bbc0c07-e52c-4de3-89cd-41f6fd3e4e57)

**Comment by MedAzizTousli** [comment]
I have the exact same issue as @RobbyCBennett.

**Comment by RobbyCBennett** [comment]
The latest fix didn't work for me either.

**Comment by Philipou** [comment]
Hello, i also have the same issue. I am on the latest ssh pre-release version and the processes just stay open and eat ram on my server is there going to be any fix ?

**Comment by FreezyEx** [comment]
Same issue for me
Remote SSH v0.113.1

**Comment by Royikiss** [comment]
fortunately, I found an issue submission with the same problem as mine. Unfortunately, it doesn't seem to be resolved? I switched to the Insiders version, which seems to be later than yours... but there is still a situation where the memory continues to increase after a period of use:
![240910_19h27m07s_screenshot](https://github.com/user-attachments/assets/fa029365-7a52-405c-9ab5-087d5dd3354e)
the following is the process after I closed vscode and used ssh to connect directly to my cloud host:
![240910_19h30m46s_screenshot](https://github.com/user-attachments/assets/70011782-a040-414a-8317-b8299cbea2be)
It seems that the related process is not automatically closed.

**Comment by jollychang** [comment]
Name: Remote - SSH
Id: ms-vscode-remote.remote-ssh
Description: Open any folder on a remote machine using SSH and take advantage of VS Code's full feature set.
Version: 0.113.1
Publisher: ms-vscode-remote
is the same issue

**Comment by xiandong79** [comment]
same issue here.

**Comment by henkedk** [comment]
+1

**Comment by dvdMucci** [comment]
+1

**Comment by Zhan-Li** [comment]
Same. Happening to two people on our teams. vscode-server constantly overloads our dev box.

**Comment by HeavyCrab** [comment]
same issue here

**Comment by GPS4PETS** [comment]
same issue here

**Comment by eferos93** [comment]
Same issue

**Comment by HagegeR** [comment]
seems like it's been back for a few weeks

**Comment by Zebnastien** [comment]
Same issue. Our servers are crashing when we ssh with VSCode after a certain amount of connections.

**Comment by dahLucky** [comment]
+1

**Comment by jcready** [comment]
@connor4312 would you be so kind as to lock this thread to contributors at least until something changes? I keep getting notifications about this issue and excitedly check for any kind of status update only to end up seeing messages from these bozos above just saying they have the same problem in a comment instead of just adding 👍 to the original post.

**Comment by Zebnastien** [comment]
You could just ask for a thumbs-up instead of being dismissive.
If you don't explain the rules, there will be bozos.

Thumbs up!

**Comment by Zebnastien** [comment]
@jcready, my bad, thanks for kindely leveling me up.

**Comment by robbycbennett** [comment]
Here's a workaround that I use on Linux and Mac SSH servers. It works for me because I don't have node running on those machines for my username.

```sh
clean() {
    pkill -2 -u $USER -f code
    pkill -2 -u $USER -f node
    pkill -2 -u $USER -f userver
}
```

**Comment by ashe0047** [comment]
Is this ever getting fixed?

**Comment by Pnlvfx** [comment]
Vscode keep releasing updates while a bug like this is still there after 1 year, isn't this a priority?

**Comment by hudynko** [comment]
This bug is not funny ... OOM kill event 3 times on my server :(

**Comment by MGAndreasen** [comment]
Imagin you have a handfull of people using the same staging server like we do.
This should be not a first priority bug to solve, but THE first priority bug to solve.

**Comment by betsch85** [comment]
after reading through PRs/issues/SO-posts for 4 hours, adding `files.watcherExclude` to every `settings.json`, adding `"watchOptions": { "excludeDirectories": ["**/node_modules/**, "/abs/path/to/node_modules/**""] } }` to every `tsconfig.json` (global, app, node) it seems like I finally found a solution: `js/ts.tsserver.watchOptions`. Even though IntelliSense  claims the only value for that key is `vscode` (to tell typescript that it should use vscode's file watcher, which also doesn't work) apparently you can give that key the same object that the `tsconfig.json` accepts. So, after putting
 ```
    "js/ts.tsserver.watchOptions": {
        "excludeDirectories": ["/abs/path/to/node_modules/**"]
    }
```
in the `settings.json` of my workspace and saving, all of the inotify watches were gone immediately. Hooray!

**Comment by SimonSiefke** [comment]
<img width="938" height="158" alt="image" src="https://github.com/user-attachments/assets/47c36ad3-ffa3-4a66-9e25-766988101938" />

Most likely, yes.  😅


<hr>

PR https://github.com/microsoft/vscode/pull/319223
